[Feature] Added Delfin UI support. Storage Summary page and widgets.

anvithks opened the pull request:
**What type of PR is this?**
/kind new feature

**What this PR does / why we need it**:
This PR adds the Delfin UI to the dashboard.
This will allow the user to use the Storage Resource Manager APIs provided by Delfin for infrastructure management and monitoring.
The PR adds the following features:
- Access Storage Summary Dashboard
- List Storage devices.
- Expand to view storage device details
- Storage Device operations
- Register Alert Source
- Sync Storage Device
- Delete Storage device
- Storage Array device tree widget
- Overview panel for quick view of storage device information
- Array Usable Capacity Widget
- Array Raw Capacity Widget
- Switch between grid view and Table view
**Which issue(s) this PR fixes**:
Fixes #370 #379 #380 #381 #382 #394

**Test Report Added?**:

/kind TESTED


**Test Report**:
Currently tested with Fake Driver support and json object from real device.

**List Storage devices**
![storage-summar-list-view](https://user-images.githubusercontent.com/19162717/92765053-9eabaa00-f3b2-11ea-90fa-37850e066f6e.png)

**Expand to show details of device**
![storage-summar-list-view-expand-details](https://user-images.githubusercontent.com/19162717/92765072-a2d7c780-f3b2-11ea-8b6e-899fb643da61.png)

**Show operations**
![storage-summar-list-view-operations-menu](https://user-images.githubusercontent.com/19162717/92765076-a4a18b00-f3b2-11ea-9b13-6e216eb68143.png)

**Confirm delete**
![storage-summar-list-view-delete-confirm](https://user-images.githubusercontent.com/19162717/92765058-a0756d80-f3b2-11ea-858f-57f3d0e44c15.png)

**Delete Success**
![storage-summar-list-view-delete-success](https://user-images.githubusercontent.com/19162717/92765065-a1a69a80-f3b2-11ea-8d06-ec883bebf48c.png)

**Register Alert Source**
![storage-summar-list-view-register-alert-source-1](https://user-images.githubusercontent.com/19162717/92765083-a53a2180-f3b2-11ea-97d0-41949884c0d6.png)
**Register Alert Source**
![storage-summar-list-view-register-alert-source-2](https://user-images.githubusercontent.com/19162717/92765088-a66b4e80-f3b2-11ea-8962-fda225493757.png)
**Register Alert Source**
![storage-summar-list-view-register-alert-source-3](https://user-images.githubusercontent.com/19162717/92765095-a79c7b80-f3b2-11ea-898e-4dcf8b0abc95.png)

**Trigger Sync**
![storage-summar-list-view-sync-request--successful](https://user-images.githubusercontent.com/19162717/92765099-a8cda880-f3b2-11ea-8846-faf55549f773.png)

**Array Usable and Raw capacity Widgets**
![storage-summary-overview-panel](https://user-images.githubusercontent.com/19162717/92931788-9c545900-f461-11ea-983c-83c6863e71ff.png)
![storage-summary-overview-panel-withwarning](https://user-images.githubusercontent.com/19162717/92931792-9f4f4980-f461-11ea-9c30-9361ba133f49.png)
![storage-summary-usable-capacity-widget](https://user-images.githubusercontent.com/19162717/92931822-aa09de80-f461-11ea-81e5-e085bb788d72.png)
![storage-summary-usable-capacity-widget-diffferent-usage](https://user-images.githubusercontent.com/19162717/92931836-aece9280-f461-11ea-9200-935fed169767.png)

**Special notes for your reviewer**:

PravinRanjan10 commented on src/app/business/delfin/storages/storages.component.html:
```diff
@@ -0,0 +1,541 @@
+                                        </span>
+                                    </div>
+                                </div>
+                                <!-- <div *ngIf="device.capacity && device.capacity.usage" class="storage-usage-bar" style="padding: 5px 25px 5px 5px;">
```
same here

anvithks replied:
Fixed.

PravinRanjan10 commented on src/app/business/delfin/storages/storages.component.html:
```diff
@@ -0,0 +1,541 @@
+                        <p-scrollPanel [style]="{width: '100%', height: '300px'}" styleClass="storage-summary-panel-2">
+                            <p-tree [value]="arrayTreeData" [loading]="loading" [style]="{width: '100%', height: '100%', 'overflow':'auto'}">
+                                <ng-template let-node pTemplate="device">
+                                    <!-- <a [routerLink]="['/storageDetails/' + node.details.id]"  (mouseenter)="showOverview($event, node, op1)" (mouseleave)="showOverview($event, node, op1)">{{node.label}}</a> -->
```
Please remove commented lines

anvithks replied:
Fixed.

PravinRanjan10 commented on src/app/business/delfin/storages/storages.component.ts:
```diff
@@ -0,0 +1,601 @@
+        });
+    }
+    getVolumesByStorage(storageId){
+        this.ds.getAllVolumes().subscribe((res)=>{
```
Doubt: we have already API for getting single volume. Then why we are calling getAllVolumes(). We may call, ex: getAllVolumesById("storageId"). This will be very fast.

anvithks replied:
Will fix this. When I started the work was not aware of the filter. Will fix this in the subsequent PRs.

anvithks replied:
Raised an issue to track this. #396

PravinRanjan10 commented on src/app/business/delfin/storages/storages.component.ts:
```diff
@@ -0,0 +1,601 @@
+    }
+
+    getPoolsByStorage(storageId){
+        this.ds.getAllStoragePools().subscribe((res)=>{
```
same here. Like volumes

anvithks replied:
Got it. Will fix this in the subsequent PRs.

kumarashit replied:
If you are fixing in subsequent PR, please open issue for this

anvithks replied:
Raised an issue to track this. #397

anvithks commented on src/app/business/delfin/storages/storages.component.ts:
```diff
@@ -0,0 +1,601 @@
+        }
+    };
+    validRule= {
+        'validIp': '([0-9]{1,3})[.]([0-9]{1,3})[.]([0-9]{1,3})[.]([0-9]{1,3})' /* Validates IPv4 address */
```
Did not quite get the issue.
Are you saying that it should match 999.999.999.999 or it will not match?

PravinRanjan10 replied:
It will match, which is not Valid IP

anvithks replied:
Got it. Will update the regex to match valid IPV4

anvithks replied:
@PravinRanjan10 Changed the regex to a better matching one. The regex can be seen in action [here](https://regex101.com/r/Ul8BDc/1)

kumarashit commented on proxy.conf.json:
```diff
@@ -1,12 +1,12 @@
 {
     "/v1beta": {
```
Generic question: Why the IPs are hardcoded? I mean what IPs are these?
Also, as in entrypoint.json, it has localhost API as DELFIN_URL, so to separate delfin n dashboard on different nodes is not possible

anvithks replied:
@kumarashit This is the file used for local development. Updated it in this PR as anyone wishing to work on delfin ui will need these changes.

kumarashit commented on src/app/business/delfin/storages/storages.component.html:
```diff
@@ -0,0 +1,528 @@
+                                    </div>
+                                </div>
+                                <div *ngIf="device.capacity && device.capacity.usage" class="storage-usage-bar" style="padding: 5px 25px 5px 5px;">
+                                    <p-progressBar [ngClass]="{'capacity-usage-normal': device.capacity.usage < 75, 
```
Can the soft and hard limit be user configurable?

anvithks replied:
Can be added as a enhancement . At the moment we can keep certain limits. Will take this up in the next set of requirments.
Will raise an issue to track this.

anvithks replied:
Raised an issue to track this. #398

kumarashit commented on src/app/business/delfin/storages/storages.component.html:
```diff
@@ -0,0 +1,528 @@
+                        <ng-template pTemplate="body" let-storage="rowData">
+                            <span class="storage-status-icon" [ngClass]="{normal: storage.status=='normal', abnormal: storage.status=='abnormal' || storage.status=='offline'}"><i class="fa fa-circle"></i></span>
+                            <span *ngIf="storage.status=='normal'">Normal</span>
+                            <span *ngIf="storage.status=='abnormal'">Abnormal</span>
```
What is this status Abnormal? I don't think we should show the status as Abnormal.. It can be "Error"

anvithks replied:
This is as per the [API spec](https://github.com/sodafoundation/delfin/blob/b0459fc0c44c13f4e4a8937a7caa22c747d6b559/openapi-spec/swagger.yaml#L85).

![image](https://user-images.githubusercontent.com/19162717/93019882-6fd24580-f5f7-11ea-8565-0911bdaf5d1b.png)

kumarashit replied:
Please check with @NajmudheenCT , but IMO you can simply do 
```
 <span *ngIf="storage.status=='Error'">Abnormal</span>
```

NajmudheenCT replied:
Checking with other stake holders , Need a unified term for 'Error' , 'Failure' , 'Unknown' etc.. any status which is not normal is mapped to 'abnormal' now..

anvithks replied:
Let me know the state machine and the values to display to the user . At the moment we have the following:
normal -> Normal
abnormal -> Abnormal
offline -> Offline

kumarashit replied:
IMHO, if !normal = Error, unless if we report the specifics

kumarashit replied:
@anvithks please open an issue to track reporting the status properly.

anvithks replied:
@NajmudheenCT @kumarashit 
Raised #400 to track this.
I would suggest getting the status codes fixed at the earliest. There will be mulitple places where the status code will need to be checked and appropriate style may be applied or operation may be allowed / disallowed.

kumarashit commented on src/app/business/delfin/storages/storages.component.html:
```diff
@@ -0,0 +1,528 @@
+                                                ...
+                                            </a>
+                                        </div>
+                                        <div class="ui-grid-col-2">
```
Is there any possibility of adding list of properties and reading them in a loop? Else, everytime a field is added, the code here need to be changed and the this file keep growing.

anvithks replied:
This may not be possible as each of the fields is a different type and there is no type identifier being returned in the API response. There are some fields which require some calculations and checks. i will explore any possibilities.

kumarashit replied:
ok

kumarashit commented on src/app/business/delfin/storages/storages.component.ts:
```diff
@@ -0,0 +1,601 @@
+        host: "Host IP",
+        port: "Port",
+        username: "Username",
+        password: "Password",
```
Why password in the description?

anvithks replied:
This is a remnant of old code. Will fix this. 
I had written Register storage as a pop-up modal and for that this object held the labels for the form. This `password` displays the label of the form .
Since then Register Storage is moved to a new page.

kumarashit replied:
Is it fixed or will be fixed in new PR?

anvithks replied:
This is pushed in the latest commit. Fixed.

kumarashit commented on src/app/business/delfin/storages/storages.component.ts:
```diff
@@ -0,0 +1,601 @@
+                })
+                msg = "<h3>Are you sure you want to delete the selected " + storages.length + " Storages?</h3><h4>[ "+ selectedNames.join(',') +" Storage(s) ]</h4>";
+            }else{
+                arr.push(storages.id)
```
Why the batch delete storage will fall in this situation? It is internal call and user directly calls this, so is this part of code required?

anvithks replied:
For deleting a storage device there are two ways. 
One is by clicking on `More` in the operations  menu and clicking on the delete button for a particular device.
The other method is by using the checkbox to select multiple devices.
Here I reuse the `batchDeleteStorages()` method in both cases. 
In case the object received is an array then pick the `id`, push into the `arr` array and display the names to the user to confirm delete.
In case it is an object and not an array then push the `id` into the `arr` array and show the appropriate message to the user.
In both cases an array of objects `arr` is sent for deletion.
If multiple selected then its an array of multiple objects else its an array of one object

kumarashit replied:
Ok

kumarashit commented on src/app/business/delfin/storages/storages.component.ts:
```diff
@@ -0,0 +1,601 @@
+            this.router.navigate(['/delfin'], {queryParams: queryParams});
+        }, (error) =>{
+            this.msgs = [];
+            this.msgs.push({severity: 'error', summary: "Error", detail:"Something went wrong. Storage device could not be registered."});
```
```suggestion
            this.msgs.push({severity: 'error', summary: "Error", detail:"Something went wrong. Storage device could not be registered for alert."});
```

anvithks replied:
The message was a typo. I will change it. The correct message should be `Alert source could not be registered.`

kumarashit commented on src/app/shared/utils/consts.ts:
```diff
@@ -57,5 +57,34 @@ export const Consts = {
+            'storagePools' : 'delfin/storage-pools',
+            'volumes' : 'delfin/volumes'
+        },
+        HOTPOT : {
```
Why HOTPOT API calls added here?

anvithks replied:
This will require checks in multiple files and that will need to be done in a different PR. I have started adding the APIs to the Consts file starting with Delfin APIs. 
There is an issue to move all API endpoints to a constants file #84 . I will reference that when i fix this. 
This is not in the scope of the current PR.

kumarashit replied:
I would suggest, let's not then mix it here. Just in case someone refers this PR in future, they should not get confused.

anvithks replied:
Fixed. Removed the references to non Delfin API endpoints. Pushed in the latest commit.

kumarashit commented on src/app/shared/utils/utils.ts:
```diff
@@ -119,6 +119,20 @@ export class Utils {
+    static formatBytes(bytes, decimals = 2) {
+        if (bytes === 0) return '0 Bytes';
+    
+        const k = 1024;
```
Can this const be kept in another constant file? Also, it need to be renamed as GiBConverter=1024 and GBConverter=1000 in constant file

anvithks replied:
Will make this change.

kumarashit replied:
Is this change done?

NajmudheenCT reviewed (changes requested):
storage sumary  report ->show serial number and sync status by defualt
storage sumary  report ->change created at to 'Discovered at ' ?
storage sumary  report ->sync_all option can be given parallel to 'Register' button ?
storage summary details report -> ID -> 'Delfin ID ' ??
storage sumary details  report ->change created at to 'Discovered at ' ?
Delete popup window :- was heading 'Delete file share' a typo ??
Trigger sync :- Do you have a logic to check whether sync is completed , I hope once you trigger sync, you should have a periodic check in loop to update sync status, exit from the loop when sync_status is 'synced'

anvithks commented:
> storage sumary report ->show serial number and sync status by defualt  

Done.
![list-storages-sync-serialnumber](https://user-images.githubusercontent.com/19162717/93083658-8ab4c080-f6b0-11ea-970f-e88c66f46a27.png)

> storage sumary report ->change created at to 'Discovered at ' ?

Changed this to `Registered at` going with the current flow as we are providing a `Register Storage` button.
![list-view-expand-details-updated](https://user-images.githubusercontent.com/19162717/93083840-d1a2b600-f6b0-11ea-9855-4d131b1eb181.png)

> storage sumary report ->sync_all option can be given parallel to 'Register' button ?

Done. Added a Sync all button.
![image](https://user-images.githubusercontent.com/19162717/93083932-f5fe9280-f6b0-11ea-8dd8-444e82e0f546.png)

![list-storages-sync-all](https://user-images.githubusercontent.com/19162717/93083878-e121ff00-f6b0-11ea-8ac9-167bc41a3ac6.png)

> storage summary details report -> ID -> 'Delfin ID ' ??

Done. 

> storage sumary details report ->change created at to 'Discovered at ' ?

Changed this to `Registered at` going with the current flow as we are providing a `Register Storage` button.

> Delete popup window :- was heading 'Delete file share' a typo ??

Thanks for pointing this out. Yes it was a copy-paste error :upside_down_face: Fixed it.
![delete-pop-up](https://user-images.githubusercontent.com/19162717/93084035-1890ab80-f6b1-11ea-8b1c-364beddee0b5.png)

> Trigger sync :- Do you have a logic to check whether sync is completed , I hope once you trigger sync, you should have a periodic check in loop to update sync status, exit from the loop when sync_status is 'synced'

The current dashbooad implementation is in a pull model. The information is pulled every page load or when the refresh button is clicked (not browser refresh, the refresh provided to invoke GET).
The user will see the latest information as per the last GET.
After the sync is triggered until the status changes to SYNCED the user will see the status as SYNCING with a spinner.
When the table is refreshed the status will be updated.
Implementing the check in a loop will present some challenges:
- Sync may take a lot of time. There is no end time and to wait for the status to be synced will not work as it may remain syncing for a while. In this case the loop will remain on.

anvithks commented:
Added Group by Model to the Storage Array Tree widget
![group-by-models](https://user-images.githubusercontent.com/19162717/93086454-c782b680-f6b4-11ea-9800-3731dc9f1735.png)

Added model overview panel.
This panel shows a quick overview by aggregating the details at the Model level
![model-overview](https://user-images.githubusercontent.com/19162717/93086510-dbc6b380-f6b4-11ea-84f1-ba8a2f5d5699.png)

kumarashit reviewed (approved):
LGTM

PravinRanjan10 commented:
![a](https://user-images.githubusercontent.com/41313813/93118644-33c5e000-f6de-11ea-946d-1c3d1090f4ab.png)
TO Remove this, please add new line at end of file.

LGTM

anvithks commented:
> ![a](https://user-images.githubusercontent.com/41313813/93118644-33c5e000-f6de-11ea-946d-1c3d1090f4ab.png)
> TO Remove this, please add new line at end of file.
> 
> LGTM

@PravinRanjan10 I am not able to see which file this line is from. 
let me know that and I will fix it.

kumarashit commented:
> > ![a](https://user-images.githubusercontent.com/41313813/93118644-33c5e000-f6de-11ea-946d-1c3d1090f4ab.png)
> > TO Remove this, please add new line at end of file.
> > LGTM
> 
> @PravinRanjan10 I am not able to see which file this line is from.
> let me know that and I will fix it.

src/app/business/delfin/delfin.service.ts for example check this file. add newline at the end

anvithks commented:
> > > ![a](https://user-images.githubusercontent.com/41313813/93118644-33c5e000-f6de-11ea-946d-1c3d1090f4ab.png)
> > > TO Remove this, please add new line at end of file.
> > > LGTM
> > 
> > 
> > @PravinRanjan10 I am not able to see which file this line is from.
> > let me know that and I will fix it.
> 
> src/app/business/delfin/delfin.service.ts for example check this file. add newline at the end

@PravinRanjan10 @kumarashit Done. Pushed in latest commit

NajmudheenCT reviewed (approved):
LGTM

anvithks commented:
@skdwriting Please review and approve.

skdwriting reviewed (approved):
LGTM
